### ISSUE TITLE
How to use both Amplify Console and CLI at the same time?

### ISSUE THREAD

**Issue by houmark** [aws-amplify/amplify-cli#1795]
**Which Category is your question related to?**

CLI and Console at the same time.

**What AWS Services are you utilizing?**

I'm running multienv, got the usual `dev` and `prod`. Project is live. To push API or function changes faster I like to use the CLI. To release in full and to deploy to production, I like to merge to my master branch and have it trigger the automatic production build. For testing major changes before deploying to production, I like to push to my develop branch and have it build also using the Console.

Here's what I'm confused about:

* The Console seems to not really care much about the overall state and just pushes according to the CloudFormation templates pulled from the branch, but could it be missing some files that are ignored in my Git?
* The CLI seems to get out of sync and not know what's the current status when Console is also deploying
* Doing `amplify env pull` does not get the CLI up to speed if I my latest deployment is with the Console. Only when I do `amplify push`, the CLI will say "No Change" for all functions and API's.

So my question is. How do I bring both in sync and do I risk any bad overwrites doing `amplify push` if I don't do `amplify env pull` after having deployed with Console?

I thought `amplify env pull` would also pull the cloud down? `amplify pull env --restore` seems a bit dangerous and I'm worried it could overwrite locally changed files?

**Comment by nino-moreton** [comment]
This has lead to us frequently doing `amplify env pull --restore` whenever a push to the amplify api fails or when we do `amplify status` and it tells has something has changed in our resources when really there weren't any (just doing `amplify env checkout [env name]` simply doesn't work for this case.

**Comment by houmark** [comment]
Yeah, I wanted to test that out more also, and close monitor which files it overwrites in my local working copy, but I seem to have constant changes that I don't want to lose, so I haven't had the opportunity to try this out that.

I think the CLI should be able to detect a more recent Console release when doing `amplify status` or at least `amplify env pull` and then do an automatic `--restore` or a more gentle one that simply gets the affected services set to "No change".

I would love clarification on this and know if something is in the works towards better sync or compatibility between the CLI and Console from the core Amplify developers.

**Comment by kaustavghosh06** [comment]
@houmark Yes, we're currently working to make the console + CLI synchronization smoother and make things which you've mentioned above simpler.

**Comment by kaustavghosh06** [comment]
@houmark 
1. Do you share the same backend infrastructure from the CLI as well as the Console?
2. If yes, `amplify env pull` should pull in the current cloud configurations into your `amplify/#current-cloud-backend` directory configurations. The `amplify evn pull --restore` command would overwrite your `amplify/backend` directory with whatever configurations you have in the cloud (stored in `amplify/#current-cloud-backend`` when you deploy it via the Amplify console.

**Comment by houmark** [comment]
Thanks for your feedback @kaustavghosh06!

Not sure what you mean with 1. But I would think my answer is yes, I have nothing custom or special running, I actually only just found out about the Console existing once I was looking to connect a real domain with my web app when I was getting close to the first release. All through development I only used the CLI and as I mentioned, I still do, as it's faster in development to switch to a terminal and do `amplify push -y` when I have done changes to for example a function. This way I am not deploying the frontend either.

For 2. I am not seeing that, but can you point me to what exactly I should expect when doing `--restore  `. This only overwrites amplify/backend and below? What if #current-cloud-backend for some reason is out of sync? I had a really weird situation trying to add Pinpoint, where it would recreate pinpoint on every single deploy in Console and at some point, I could not deploy as I had reached a policy limit of 10. I found the policies and deleted them manually and deleted pinpoint, but never really found out what happened. I still have one pinpoint hanging around in my dev environment, but production is clean and if I delete that one then I will recreate. But I do not see pinpoint in `amplify status`. Something got corrupted and debugging and trying to clean up is really hard. If this happened due to some kind of confusion between Console and the CLI or it was a CLI bug I am not sure, but I'd love to get things cleaned up.

In any case, my current take away is that if I have a fully clean local working copy of my Git with no local changes, and I still see updates in the CLI on `amplify status` it should be safe to do `amplify env pull --restore`?

Something like a more in-depth check in the cloud when doing `amplify status` or `amplify checkout env` with an automatic pull of cloud files and overwrite locally, if the local folder/Git is clean and with a CLI warning/question to update/overwrite local files with cloud files, if Git is not clean, would be nice.

**Comment by swaminator** [comment]
@hourmark
Some questions:
1. When you connected your app in the console did you pick an existing environment or did you create a new one?
2. We are working on providing more insight in the console wrt the backend. A developer should be able to:
- View resources and environments created in the CLI from the Console.
- Point the CLI created backend to a specific frontend in the Console.
- Delete resources and environments from the Console and have a way to reflect the changes in the CLI
- Update resources in the Console and have a way to pull those changes to the CLI.

What else would you like to see?

**Comment by houmark** [comment]
@swaminator 

1. I picked an existing one (well, two, I connected both dev and prod which existed already)

Thanks for the update on your vision on the Console. It has a lot to wish for still, but it's a good start.

I would like to see:

1. Ability to do sub sub domains for a domain, even while keeping the root domain out of play. Our setup is site.com with our main site. members.site.com which is the Amplify React SPA. And for testing/staging we would like to have for example testing.member.site.com

**Comment by houmark** [comment]
@swaminator 

1. I picked an existing one (well, two, I connected both dev and prod which existed already)

Thanks for the update on your vision on the Console. It has a lot to wish for still, but it's a good start.

Besides perfect sync between CLI and Console, I would like to see:

1. Ability to do sub sub domains for a domain, even while keeping the root domain out of play. Our setup is site.com with our main site. members.site.com which is the Amplify React SPA. And for testing/staging we would like to have for example testing.member.site.com

2. Some way to not deploy frontend of backend, so we can quicker push out changes in the Console without having to wait for either of the two slower steps.

3. Outbound webhooks to trigger after a deployment has fully finished and is verified. Or at least a webhook no matter what happens with the result. This could be useful to somehow trigger a subscription for the user to update the site "due to a new version available".

I'm still new around here, so I may come back to add more comments if anything else comes to mind, but for now, thanks for listening and being interested in your users feedback, something great is in the making with the Amplify platform, can't wait to see where all this is in one year from now.

**Comment by swaminator** [comment]
@houmark  
1. Checkout https://docs.aws.amazon.com/amplify/latest/userguide/custom-domains.html#custom-domain-subdomains
2. Can you explain a bit more?
3. You can add Email notifications in the Console. This actually creates an SNS topic which you can use however you want (setup Slack notification as an example)

Thanks for the feedback.

**Comment by houmark** [comment]
@swaminator Thanks for your feedback!

1. Hmm, I thought I had read that but couldn't find it again. I guess what confused me was that it's kinda warning when you have the main domain abc.com and type in sub.abc.com as the root domain, so it's not very intuitive. I guess a text like "want to create sub sub domains? Check out the documentation" with a link to that page would be helpful. Or maybe I'm the only one that couldn't connect the dots here.

2. When I push one commit to the develop branch, it will run both backend and frontend in full. Maybe it could auto detect that the backend has no changes internally using "amplify status" and just jump that step for a faster frontend deployment (think bug in production in the frontend code, push a fix, but have to wait for the backend to deploy which can be some heavy 5+ minutes). Manual triggers for deciding which parts would be nice also. It could for example be based on Git commit messages and the content thereof. For example, if you put [deploy:frontend] it would only deploy the frontend (this idea is based on another Git hosted platform I am using). 

3. Nice to know. That may work for our need, I'll check it out.

In the meantime, today I had a clean working copy in my dev branch for the first time in a good period and I had deployed with the Console, so doing `amplify status` resulted in many changed resources (that was already deployed through the Console). I did a `amplify env pull --restore` which resulted in some new CloudFormation files locally. Not a biggie and since the Console cannot (and should not!) write to the Git branch, these files need to be committed (which then will result in a new deployment in Console, that I believe it could auto ignore with some more pre-flight checks on the changes committed).

The main take away from this, in relation to confusion for me as a developer, was this error on my next CLI push right after showing the current status:

```
Could not read cloudformation template at path: 

`/Site/project/amplify/backend/api/myapi/build/cloudformation-template.json`
```

Does not seem to affect anything and I have seen this once before, but it eventually went away (not sure what made it go away though). Suggestions on that one?
